File: frontend/frontend/src/domain/contrast.ts
/**
 * WCAG color-contrast math.
 *
 * Implements the algorithm from
 * https://www.w3.org/TR/WCAG21/#dfn-relative-luminance and the contrast ratio
 * formula in §1.4.3.
 */

export interface ContrastResult {
  ratio: number;
  /** AA pass for normal text (≥ 4.5:1). */
  aaNormal: boolean;
  /** AA pass for large text (≥ 3:1). */
  aaLarge: boolean;
  /** AAA pass for normal text (≥ 7:1). */
  aaaNormal: boolean;
  /** AAA pass for large text (≥ 4.5:1). */
  aaaLarge: boolean;
  /** AA pass for non-text UI components (≥ 3:1). */
  aaUiComponents: boolean;
}

/**
 * Parse a #RRGGBB or #RGB hex color into [r, g, b] integers in 0-255.
 * Returns null on invalid input.
 */
export function parseHex(input: string): [number, number, number] | null {
  if (!input) return null;
  const trimmed = input.trim().replace(/^#/, "");
  if (!/^[0-9a-fA-F]+$/.test(trimmed)) return null;
  let hex = trimmed;
  if (hex.length === 3) {
    hex = hex
      .split("")
      .map((c) => c + c)
      .join("");
  }
  if (hex.length !== 6) return null;
  const r = parseInt(hex.slice(0, 2), 16);
  const g = parseInt(hex.slice(2, 4), 16);
  const b = parseInt(hex.slice(4, 6), 16);
  return [r, g, b];
}

function _channelLum(c: number): number {
  const v = c / 255;
  return v <= 0.03928 ? v / 12.92 : Math.pow((v + 0.055) / 1.055, 2.4);
}

export function relativeLuminance([r, g, b]: [number, number, number]): number {
  return 0.2126 * _channelLum(r) + 0.7152 * _channelLum(g) + 0.0722 * _channelLum(b);
}

export function contrastRatio(
  fg: [number, number, number],
  bg: [number, number, number],
): number {
  const lum1 = relativeLuminance(fg);
  const lum2 = relativeLuminance(bg);
  const [light, dark] = lum1 > lum2 ? [lum1, lum2] : [lum2, lum1];
  return (light + 0.05) / (dark + 0.05);
}

type RGB = [number, number, number];

function _blend(a: RGB, b: RGB, t: number): RGB {
  return [
    Math.round(a[0] + (b[0] - a[0]) * t),
    Math.round(a[1] + (b[1] - a[1]) * t),
    Math.round(a[2] + (b[2] - a[2]) * t),
  ];
}

function _toHex([r, g, b]: RGB): string {
  return (
    "#" +
    [r, g, b]
      .map((x) => Math.max(0, Math.min(255, x)).toString(16).padStart(2, "0"))
      .join("")
      .toUpperCase()
  );
}

/**
 * Find the colour CLOSEST to ``color`` (smallest perceptual nudge) that, paired
 * with ``against``, reaches ``target`` contrast. Scans toward black and toward
 * white — a fine linear scan, not a binary search, because the ratio is U-shaped
 * (non-monotonic) when ``color`` and ``against`` straddle each other in
 * luminance. ``adjustingForeground`` controls which side of the ratio ``color``
 * sits on. Returns null when even pure black/white can't reach the target.
 */
function _nearestPassing(
  color: RGB,
  against: RGB,
  target: number,
  adjustingForeground: boolean,
): { hex: string; ratio: number } | null {
  const extremes: RGB[] = [
    [0, 0, 0],
    [255, 255, 255],
  ];
  let best: { hex: string; ratio: number; t: number } | null = null;
  for (const extreme of extremes) {
    for (let i = 1; i <= 100; i += 1) {
      const t = i / 100;
      const c = _blend(color, extreme, t);
      const ratio = adjustingForeground ? contrastRatio(c, against) : contrastRatio(against, c);
      if (ratio >= target) {
        if (!best || t < best.t) best = { hex: _toHex(c), ratio, t };
        break; // first passing step in this direction is the closest for it
      }
    }
  }
  return best ? { hex: best.hex, ratio: best.ratio } : null;
}

/**
 * Suggest the nearest passing colours for a failing pair: keep one channel and
 * nudge the other (and vice-versa) just enough to clear ``target`` (AA normal
 * = 4.5 by default). Either field is null if that side can't reach the target.
 */
export function suggestPassing(
  fgHex: string,
  bgHex: string,
  target = 4.5,
): {
  foreground: { hex: string; ratio: number } | null;
  background: { hex: string; ratio: number } | null;
} {
  const fg = parseHex(fgHex);
  const bg = parseHex(bgHex);
  if (!fg || !bg) return { foreground: null, background: null };
  return {
    foreground: _nearestPassing(fg, bg, target, true),
    background: _nearestPassing(bg, fg, target, false),
  };
}

export function evaluate(
  fgHex: string,
  bgHex: string,
): { ok: true; result: ContrastResult } | { ok: false; reason: string } {
  const fg = parseHex(fgHex);
  if (!fg) return { ok: false, reason: `Foreground "${fgHex}" is not a valid hex color.` };
  const bg = parseHex(bgHex);
  if (!bg) return { ok: false, reason: `Background "${bgHex}" is not a valid hex color.` };
  const ratio = contrastRatio(fg, bg);
  return {
    ok: true,
    result: {
      ratio,
      aaNormal: ratio >= 4.5,
      aaLarge: ratio >= 3,
      aaaNormal: ratio >= 7,
      aaaLarge: ratio >= 4.5,
      aaUiComponents: ratio >= 3,
    },
  };
}

File: backend/app/analyzers/contrast.py
"""WCAG 2.1 colour-contrast math (SC 1.4.3 / 1.4.11).

Pure functions — no document model — so they are trivially unit-testable and
reusable. Implements the exact WCAG relative-luminance + contrast-ratio formula.
"""

from __future__ import annotations

from typing import Optional, Tuple

RGB = Tuple[int, int, int]

# WCAG AA thresholds.
AA_NORMAL = 4.5
AA_LARGE = 3.0


def parse_hex(value: object) -> Optional[RGB]:
    """Parse ``#RRGGBB`` / ``RRGGBB`` / ``RGB`` into an (r, g, b) tuple, or None."""
    if value is None:
        return None
    s = str(value).strip().lstrip("#")
    if len(s) == 3:
        s = "".join(ch * 2 for ch in s)
    if len(s) != 6:
        return None
    try:
        return (int(s[0:2], 16), int(s[2:4], 16), int(s[4:6], 16))
    except ValueError:
        return None


def _channel_luminance(c: float) -> float:
    # c is 0..1
    return c / 12.92 if c <= 0.03928 else ((c + 0.055) / 1.055) ** 2.4


def relative_luminance(rgb: RGB) -> float:
    """WCAG relative luminance of an sRGB colour."""
    r, g, b = (channel / 255.0 for channel in rgb)
    return 0.2126 * _channel_luminance(r) + 0.7152 * _channel_luminance(g) + 0.0722 * _channel_luminance(b)


def contrast_ratio(fg: object, bg: object) -> Optional[float]:
    """Contrast ratio (1..21) between two colours, or None if unparseable."""
    f = parse_hex(fg)
    b = parse_hex(bg)
    if f is None or b is None:
        return None
    l1 = relative_luminance(f)
    l2 = relative_luminance(b)
    hi, lo = (l1, l2) if l1 >= l2 else (l2, l1)
    return (hi + 0.05) / (lo + 0.05)


def is_large_text(size_pt: Optional[float], bold: bool) -> bool:
    """WCAG "large text": >= 18pt, or >= 14pt bold."""
    if size_pt is None:
        return False
    try:
        pt = float(size_pt)
    except (TypeError, ValueError):
        return False
    return pt >= 18.0 or (bold and pt >= 14.0)


def required_ratio(size_pt: Optional[float], bold: bool = False) -> float:
    """The AA minimum ratio for text of this size/weight (3.0 large, else 4.5)."""
    return AA_LARGE if is_large_text(size_pt, bold) else AA_NORMAL


def _to_hex(rgb: RGB) -> str:
    return "".join(f"{max(0, min(255, int(round(c)))):02X}" for c in rgb)


def _toward(src: RGB, target: RGB, t: float) -> RGB:
    return (
        src[0] + (target[0] - src[0]) * t,
        src[1] + (target[1] - src[1]) * t,
        src[2] + (target[2] - src[2]) * t,
    )


def suggest_passing_fg(fg: object, bg: object, required: float = AA_NORMAL) -> Optional[str]:
    """The nearest foreground colour to ``fg`` that meets ``required`` contrast
    against ``bg``, as ``RRGGBB`` (no ``#``), or None.

    We never change the background (recolouring a page is a bigger design
    decision than nudging text), and we change the text colour as little as
    possible: scan from ``fg`` toward black and toward white, take the first
    passing colour found along each direction (the nearest in that direction),
    and return whichever of the two is the smaller change. Returns None if
    ``fg``/``bg`` are unparseable or ``fg`` already passes (nothing to suggest)."""
    f = parse_hex(fg)
    b = parse_hex(bg)
    if f is None or b is None:
        return None
    current = contrast_ratio(fg, bg)
    if current is not None and current >= required:
        return None

    best: Optional[Tuple[float, str]] = None  # (squared-distance, hex)
    steps = 64
    for target in ((0, 0, 0), (255, 255, 255)):
        for i in range(1, steps + 1):
            cand = _toward(f, target, i / steps)
            cand_hex = _to_hex(cand)
            r = contrast_ratio(cand_hex, bg)
            if r is not None and r >= required:
                dist = sum((cand[k] - f[k]) ** 2 for k in range(3))
                if best is None or dist < best[0]:
                    best = (dist, cand_hex)
                break  # first passing along this direction is the nearest in it
    return best[1] if best else None


__all__ = [
    "AA_NORMAL",
    "AA_LARGE",
    "parse_hex",
    "relative_luminance",
    "contrast_ratio",
    "is_large_text",
    "required_ratio",
    "suggest_passing_fg",
]
